### ISSUE TITLE
TS ^4 does not recognize array methods on (T[] | T[][]) as callable

### ISSUE THREAD

**Issue by dkamyshov** [microsoft/TypeScript#40157]
**TypeScript Version:**  4.0.2, 4.1.0-dev.20200820

**Search Terms:** union, array methods

**Code**

The following example uses `every` method, but the bug affects all other methods as well (`map`, `reduce`, etc.).

```ts
const a: string[] | string[][] = [];

a.every(b => console.log(b)); // TS-2349: This expression is not callable.

// Attempt №1: suppress TS-7006
a.every((b: string | string[]) => console.log(b)); // TS-2349: This expression is not callable.

// Attempt №2: rewrite as typeguard (as in https://github.com/microsoft/TypeScript/blob/v4.0-beta/lib/lib.es5.d.ts#L1319)
a.every((b: string | string[]): b is string => typeof b === 'string'); // TS-2349: This expression is not callable.
```

**Expected behavior:** No errors, code should compile.

**Actual behavior:** Error `TS-2349: This expression is not callable.`

**Playground Link:** https://www.typescriptlang.org/play?ts=4.1.0-dev.20200819#code/MYewdgzgLgBAhgLhtATgSzAcwNoF0YA+yU6WeeMAvDHgNwBQ9cAdAKYBurKAngBQBGVAHwxQkEABtWzCSEwCAlAoZM2nHrwFJUGTIWKkcuBcNHgIk6bPn8lKlhy58tB3fp1ljSQWgiusplDcAA6sIABmMIKUMTAA5B6Yccr0QA

**Related Issues:** I didn't find anything similar.

**Comment by dkamyshov** [comment]
The error disappears if the array is initialized with something inside.

```ts
const a: string[] | string[][] = [];

a.every(() => {}); // TS-2349: This expression is not callable.

const b: string[] | string[][] = [[]];

b.every(() => {}); // ok

const c: string[] | string[][] = ['c'];

c.every(() => {}); // ok

const f = (d: string[] | string[][]) => {
    d.every(() => {}); // TS-2349: This expression is not callable.
}

```

https://www.typescriptlang.org/play?ts=4.1.0-dev.20200819#code/MYewdgzgLgBAhgLhtATgSzAcwNoF0YA+yU6WeeMAvDHgNwBQ9cAdAKYBurKAngBS8BKKgD4YAbwC+AhvVCRYAIySoMOfERVlcFatjy4ZCtpx78hlUZOmM50GMGUlVFDU606aAcmCeDN41x8giLiUjK2sABmVDC8ACaOpGqExEnkuOaW9DA5MHEBpsEWodYS9EA

**Comment by dkamyshov** [comment]
UPD: methods, that do not have generic overloads (`some`, `forEach`) are not affected.

**Comment by weswigham** [comment]
This is working as intended (or at least a current design limitation), if a breaking change in 4.0 due to a `lib` update - https://github.com/microsoft/TypeScript/pull/38200 added a second, generic overload to `.every`, making it so we couldn't resolve a single combined signature for the union (and so it now behaves as `.map` and friends).

Changing the types you annotate on the argument to `.every` aren't going to make `.every` itself callable. The issue is that the union of the two array types contains irreconcilable overload lists for `.every` (we'd have to create a power set of signatures and resolve them as a new overload list, while also unifying generics in similar positions, and even then it's not necessarily correct) - you need to cast the `.every` call itself (which is going to break any type guards you try to apply, mind you).

**Comment by weswigham** [comment]
@RyanCavanaugh up to you - this break should possibly be noted in the release breaking changes, if we want to keep it.

**Comment by pleunv** [comment]
So is there currently any _proper_ workaround? I can't seem to get around the error.
